### PR TITLE
Use the OPRF DeriveKeyPair function and update OPRF usage

### LINKED ISSUE #241: Tighten up the OPRF abstraction
The OPRF abstraction currently (and intentionally) leaks some underlying implementation details. In particular, it leaks the fact that it requires a prime-order group. Should we want to augment OPAQUE to make it PQ-friendly, PQ-friendly OPRFs, or even threshold variants, should ideally work here. We should try to improve this abstraction so as to accommodate other types of primitives.

### LINKED ISSUE #328: Adopt VOPRF DeriveKeyPair
... and fix serialization stuff, as that's no longer part of the VOPRF abstraction.

### PR DESCRIPTION
**Pull request by chris-wood**
This change uses the DeriveKeyPair function from the OPRF spec for both the OPRF key generation and authentication key generation. It also adds serialization steps to the pseudocode and reference implementation, as that is no longer handled by the OPRF API.

Closes #328.
Closes #241.

### CODE REVIEW

**Comment by chris-wood** [comment]
@stef I reverted the change in [eae5040](https://github.com/cfrg/draft-irtf-cfrg-opaque/pull/329/commits/eae5040eec78f69eaf77a3e7b59900e40f1899ef).

File: poc/test_opaque_ake.sage [hunk 1]
```diff
@@ -33,6 +33,7 @@ def test_core_protocol_serialization():
 
     config = default_opaque_configuration 
     group = config.group
+    (_, _) = group.key_gen()
```
Comment (bytemare) [inline poc/test_opaque_ake.sage:36]: what does that do?

File: draft-irtf-cfrg-opaque.md [hunk 2]
```diff
@@ -881,12 +874,19 @@ Input:
...
+Exceptions:
+- DeserializeError, when OPRF element deserialization fails.
+
 def CreateRegistrationResponse(request, server_public_key,
                                credential_identifier, oprf_seed):
   seed = Expand(oprf_seed, concat(credential_identifier, "OprfKey"), Nseed)
-  (oprf_key, _) = DeriveKeyPair(seed)
-  Z = Evaluate(oprf_key, request.data, nil)
-  Create RegistrationResponse response with (Z, server_public_key)
+  (oprf_key, _) = DeriveKeyPair(seed, "OPAQUE")
```
Comment (bytemare) [inline draft-irtf-cfrg-opaque.md:883]:
```suggestion
  (oprf_key, _) = DeriveKeyPair(seed, "OPAQUE-DeriveKeyPair")
```

File: poc/test_opaque_ake.sage (hunk 1 above)
Comment (chris-wood) [reply poc/test_opaque_ake.sage:36]: It fixes the deterministic test vector issue @stef pointed out. I'll address this (by removing it) in a future change that _does not_ make any protocol changes.

**Review by bytemare — CHANGES_REQUESTED** [review]

File: draft-irtf-cfrg-opaque.md [hunk 3]
```diff
@@ -1179,18 +1186,25 @@ Input:
...
 
+Exceptions:
+- DeserializeError, when OPRF element deserialization fails.
+
 def CreateCredentialResponse(request, server_public_key, record,
                              credential_identifier, oprf_seed):
   seed = Expand(oprf_seed, concat(credential_identifier, "OprfKey"), Nok)
-  (oprf_key, _) = DeriveKeyPair(seed)
-  Z = Evaluate(oprf_key, request.data, nil)
+  (oprf_key, _) = DeriveKeyPair(seed, "OPAQUE")
```
Comment (bytemare) [inline draft-irtf-cfrg-opaque.md:1195]:
```suggestion
  (oprf_key, _) = DeriveKeyPair(seed, "OPAQUE-DeriveKeyPair")
```

**Review by bytemare — CHANGES_REQUESTED** [review]

**Comment by chris-wood** [comment]
@stef are you able to verify the test vectors in this PR?

**Review by bytemare — APPROVED** [review]

**Comment by stef** [comment]
with pleasure!

**Comment by bytemare** [comment]
lgtm

**Comment by bytemare** [comment]
Does this close #298 ?

**Comment by chris-wood** [comment]
> Does this close #298 ?

I don't think so. I'll address those in a followup change.

**Comment by stef** [comment]
i checked, and updated my implementation. lgtm.

**Comment by chris-wood** [comment]
Excellent! Merging.

**Pull request merged**